### ISSUE TITLE
WebSockets Next: performance improvements

### ISSUE THREAD

**Issue by mkouba** [quarkusio/quarkus#39148]
### Description

Follow-up of https://github.com/quarkusio/quarkus/pull/39142.

### Implementation ideas

- use an `AtomicLongFieldUpdater` in the `ConcurrencyLimiter`: https://github.com/quarkusio/quarkus/pull/39142#discussion_r1510855917

**Comment by mkouba** [comment]
The https://github.com/quarkusio/quarkus/pull/40183 is related to this issue.

**Comment by mkouba** [comment]
For this one, we'll need some benchmarks (executed automatically). Ideally, to compare WS next with `quarkus-websockets` and pure Vert.x.

CC @franz1981

**Comment by franz1981** [comment]
I'm not aware of websocket benchmark suites sadly...

**Comment by mkouba** [comment]
> I'm not aware of websocket benchmark suites sadly...

Me neither. We'll need to come with something... ;-)

**Comment by franz1981** [comment]
Let's talk this week in a call and we can think about something

**Comment by mkouba** [comment]
NOTE: This extension contains a bunch of lambdas. We should consider rewriting those lambdas to anonymous classes.

**Comment by cescoffier** [comment]
We need to think about scenarios to test the performances. Response time is not a meaningful metric. The number of messages and connections are more sensitive in this case. (Of course, memory is important too).

**Comment by franz1981** [comment]
Yeah, although if you can achieve 10K msg/sec with a single one with 10 seconds of latency - is something you wish to know.

I have contacted the Jetty team (Simone Bordet) - and they have rolled out a coordination omission free distributed (if required) load generator for websocket; let's see what we can do in Hyperfoil or by reusing such, which is used for this exact purpose

**Comment by cescoffier** [comment]
oh, that's would we good!

**Comment by franz1981** [comment]
And clearly it is not covering websocket, see https://github.com/jetty-project/jetty-load-generator

Which means that we should prioritize supporting websockets for Hyperfoil or find a different benchmarking tool (which is coordinated-omission free - not easy)

**Comment by mkouba** [comment]
> Which means that we should prioritize supporting websockets for Hyperfoil or find a different benchmarking tool (which is coordinated-omission free - not easy)

For load tests where we don't care about coordinated-omission and throughput, we could try to use the [Gatling WebSocket protocol](https://docs.gatling.io/reference/script/protocols/websocket/) or even a simple Vert.x client.

**Comment by cescoffier** [comment]
I used Gatling in the past.

**Comment by mkouba** [comment]
So I started to play with a simple Vertx client, Gatling, etc. here: https://github.com/mkouba/ws-next-perf.

And it seems that at moderate load the performance of `quarkus-websockets` and `quarkus-websockets-next` is more or less the same. However under heavy load (in my test it was 10.000 concurrent users sending 1000 and receiving 1000 messages) the performance of `quarkus-websockets` degrades significantly. I did some CPU profiling and I didn't find anything obvious in the WS next code.

Apparently the biggest problem is switching to a worker thread because the tested `@OnTextMessage` callback has a blocking signature. If we switch to `Uni<String>` (i.e. callback executed on the event loop) then the performance is significantly better but still not better than the legacy extension. However, the blocking signature is probably what most users will use anyway...

@franz1981 Could you pls take a look at the attached flamegraph?

[cpu-profile.html.zip](https://github.com/user-attachments/files/17275536/cpu-profile.html.zip)

**Comment by franz1981** [comment]
I see there's a fun problem with `synchronizer`s, which can impact pretty bad scalability and perf i.e. `io/vertx/core/http/impl/ServerWebSocketImpl.tryHandshake` (and `io/vertx/core/net/impl/ConnectionBase.queueForWrite` as well) which is protecting the handshake via a `synchronized` guard.
You can confirm this by collecting profiling data using `-e lock -t` (add t as well to see which threads are competing to enter in the lock).
The suggestion here is to have less worker threads which compete among each other(s), but they will likely still compete vs the I/O threads (or not - we need the profiling via `-e lock -t`).
I believe the check performed on `io/vertx/core/http/impl/ServerWebSocketImpl.tryHandshake` can be improved via some volatile guard - and avoid it - and we will fly.
I can create a specific vertx microbenchmark  for this in vertx itself


**important note**:
I'm at devoxx and I didn't yet look at the bench itself but 2 things after looking at the data:
1. collect profoling data after waiting a bit before warming up of the application complete: I can see C2 frames meaning that compilation is still going on (after ~10K or less invocations thing will smooth out)
2. it looks that it is intentionally a cpu bound computation: is it what we expect? I would, instead, add a parametrized fake blocking call `(Thread::sleep(configuredFakeBlockingWork)` to perform some really blocking behaviour, when we run things on the worker thread pool - this will make more realistic. 

The last point is very key to understand that if users are making use of the worker thread pool they are supposed to perform blocking operations (in the form of 10/100 ms work each), this will guarantee 2 effects:
1. less contention over `synchronized` part - likely
2. less "oversubscription" of worker threads vs the available cores i.e. worker threads pool have Math::min(cores * 8, 200) threads - and it means they need to interleave to make progress..if you make them cpu bound performing little non-blocking work, this can stress a LOT other OS mechanism which won't be stressed in the real world

As usual, I love you're so proactive and quick to react @mkouba thanks again for taking both time and effort for the test + collecting data: this will make so much easier for me to help! ❤

**Comment by franz1981** [comment]
In addition; this is another low hanging fruit I can help with:
![Image](https://github.com/user-attachments/assets/f4fa9b0a-662b-4aec-bd25-b91c5be3f941)

https://github.com/eclipse-vertx/vert.x/blob/916ae9911dbb2a8cf818eee6b5390f62f37fce00/vertx-core/src/main/java/io/vertx/core/http/impl/WebSocketImplBase.java#L478-L493

But gotta better check: I'm adding this note for myself of the future

**Comment by mkouba** [comment]
> I see there's a fun problem with `synchronizer`s, which can impact pretty bad scalability and perf (and RSS because "inflated" monitors increase RSS on Hotspot) i.e. `io/vertx/core/http/impl/ServerWebSocketImpl.tryHandshake` (and `io/vertx/core/net/impl/ConnectionBase.queueForWrite` as well) which is protecting the handshake via a `synchronized` guard.

Yes, I noticed this part as well.

> I can create a specific vertx microbenchmark for this in vertx itself

That would be great.

> 1. collect profoling data after waiting a bit before warming up of the application complete: I can see C2 frames meaning that compilation is still going on (after ~10K or less invocations thing will smooth out)

[cpu-profile.html_02.zip](https://github.com/user-attachments/files/17276311/cpu-profile.html_02.zip)

> 2\. it looks that it is intentionally a cpu bound computation: is it what we expect? I would, instead, add a parametrized fake blocking call `(Thread::sleep(configuredFakeBlockingWork)` to perform some really blocking behaviour, when we run things on the worker thread pool - this will make more realistic.

It depends. I don't think that all callbacks with a blocking signature will execute code that would block the thread. But for sure, we need more scenarios that would cover all common use cases. Currently, we only call `String.toLowerCase()` 🤷 .

Thanks Franz!

**Comment by mkouba** [comment]
FYI I've just noticed the following sentence in the javadoc of `io.vertx.core.http.impl.ServerWebSocketImpl`: _"This class is optimised for performance when used on the same event loop. However it can be used safely from other threads."_.

And also _"The internal state is protected using the synchronized keyword. If always used on the same event loop, then we benefit from biased locking which makes the overhead of synchronized near zero."_.

So obviously, it's not optimized for the blocking use case ;-).

**Comment by franz1981** [comment]
@mkouba Yep and ideally this could be improved  on Vertx 5, but there is still some low hanging fruit on Vertx 4 - which we can easily explore if is worthy i.e. https://github.com/franz1981/vert.x/commit/3ca72f8ee5aaaa3aebd06791a7a76c60c99a5223
If you want to try this or apply this commit to the right vertx branch you could give it a shot in your benchmark

what is doing is fairly simple, and is based on the analysis I've performed for https://github.com/franz1981/java-puzzles/blob/583d468a58a6ecaa5e7c7c300895392638f688dd/src/main/java/red/hat/puzzles/concurrent/LockCoarsening.java#L76-L85 which were the motivation behind the vertx 5 changes in this regard ie recursive/nested synchronized over the same field (connection, here) just keep on getting a performance hit for each additional nested level. In the mentioned commit I just taken care to remove the most of the nested synchronized I have found

**Comment by vietj** [comment]
FYI : this part in Vertx 5 has been rewritten, so this analysis does not hold for it

**Comment by mkouba** [comment]
> If you want to try this or apply this commit to the right vertx branch you could give it a shot in your benchmark

Unfortunately, it does not seem to be an easy task to switch the `vertx-core` version used in Quarkus. You cannot simply change the `vertx.version` in the BOM because it's used for other Vert.x dependencies (`vertx-web`, etc.). You cannot set an explicit version the `quarkus-vertx` runtime because you get dependency convergence errors. 

@cescoffier @vietj Any tip how to try this out?

**Comment by mkouba** [comment]
> FYI : this part in Vertx 5 has been rewritten, so this analysis does not hold for it

Hey Julien, do you have some benchmarks in Vert.x to test the performance of WebSockets server/client?

**Comment by mkouba** [comment]
> > Unfortunately, it does not seem to be an easy task to switch the vertx-core version used in Quarkus.
> 
> What I would do is to cherry pick the commit to the right vertx tag, use mvn install and either replace the jar in the lib of quarkus or hope that the local mvn repo will do the right thing(TM)

Ah, ofc. This worked. And quick and dirty results seem to be much better, comparable to `quarkus-websockets`.

**Comment by franz1981** [comment]
@mkouba ok so this seems a painless change if @vietj and @cescoffier agreed and you see benefits.
I spent some time analysing the weird synchronised behaviour with the vertx code pattern so, sadly, these "workarounds" can be very effective

**Comment by cescoffier** [comment]
Do you have a link to the commit to cherry-pick?

**Comment by mkouba** [comment]
> Do you have a link to the commit to cherry-pick?

@cescoffier https://github.com/franz1981/vert.x/commit/3ca72f8ee5aaaa3aebd06791a7a76c60c99a5223

**Comment by cescoffier** [comment]
The committee looks good. It avoids entering synchronized blocks.

I'm not sure of the various assertions. 

Let's see what @vietj says.

**Comment by mkouba** [comment]
> The committee looks good. It avoids entering synchronized blocks.
> 
> I'm not sure of the various assertions.
> 
> Let's see what [@vietj](https://github.com/vietj) says.

@cescoffier What committee? 😆

**Comment by franz1981** [comment]
Yep @cescoffier the checks on asserts should be enable on both quarkus and vertx maven surefire tests to make use the new methods are not misused while still not impacting the hot path at runtime (asserts are fully removed)

**Comment by franz1981** [comment]
I have created https://github.com/franz1981/vert.x/commit/9a0f5168bec041ba66811e82867e389a96f84449 to fix the buffer problem saw dew comments earlier too

**Comment by mkouba** [comment]
FYI I'm working on a pull request to disable CDI request context activation for endpoint callbacks unless really needed, i.e. an endpoint has a `@RequestScoped` dependency or is secured.

**Comment by franz1981** [comment]
@mkouba 

> we still don' have a proper benchmark.

the one you did is a good starting point and indeed I got a request for you...re 
https://github.com/quarkusio/quarkus/pull/45185
I expect that if our websocket support is on top of exiting http vertx servers, it should be able to popup in your benchmark if we could enable TLS on it - wdyt?
